lib/PerlIO.pm

```diff
@@ -154,24 +154,120 @@ removed normally.
-upgraded encoding of characters, which is approximately UTF-8 on ASCII
-machines, but UTF-EBCDIC on EBCDIC machines.  This allows any character
-Perl can represent to be read from or written to the stream.
-
-This layer (which actually sets a flag on the preceding layer, and is
-implicitly set by any C<:encoding> layer) does not translate or validate
-byte sequences.  It instead indicates that the byte stream will have been
-arranged by other layers to be provided in Perl's internal upgraded
-encoding, which Perl code (and correctly written XS code) will interpret
-as decoded Unicode characters.
-
-B<CAUTION>: Do not use this layer to translate from UTF-8 bytes, as
-invalid UTF-8 or binary data will result in malformed Perl strings.  It is
-unlikely to produce invalid UTF-8 when used for output, though it will
-instead produce UTF-EBCDIC on EBCDIC systems.  The C<:encoding(UTF-8)>
-layer (hyphen is significant) is preferred as it will ensure translation
-between valid UTF-8 bytes and valid Unicode characters.
+Declares that the stream accepts UTF-8. This allows any character perl can
+represent to be read from or written to the stream. The UTF-8 encoding
+is chosen to render simple text parts (i.e.  non-accented letters,
+digits and common punctuation) human readable in the encoded file.
+
+Here is how to write your native data out using UTF-8 (or UTF-EBCDIC)
+and then read it back in.
+
+       open(my $fh, ">:utf8", "data.utf");
+       print $fh $out;
+       close($fh);
+
+       open(my $fh, "<:utf8", "data.utf");
+       $in = <$fh>;
+       close($fh);
+
+Note that before perl 5.36, this layer did not validate byte sequences.
+
+From 5.36 C<:utf8> accepts parameters that control what errors are
+allowed and how errors are handled.  These are ignored in older
+versions of perl.
+
+If you allow encoding errors on the input stream they are always
+converted to the Unicode REPLACEMENT CHARACTER.
+
+Options that control allowed errors:
+
+=over
+
+=item *
+
+C<allow_aurrogates> - allow surrogates in the input stream.
+
+=item *
+
+C<allow_noncharacters> - allows non-characters, including the code
+points from U+FDD0 through U+FDEF and the U+XFFFE and U+XFFFF code
+points at the end of each plane.
+
+=item *
+
+C<allow_super> - allow code points beyond the Unicode legal max
+(U+10FFFF), using perl's extended UTF-8 encoding.
+
+=item *
+
+C<allow_nonshortest> - allow non-shortest or overlongs in the input
+stream.  This is an encoding error and is replaced with a replacement
+character if allowed.
+
+=item *
+
+C<allow_continuation> - allow UTF-8 continuation bytes out of
+sequence.  This is an encoding error and is replaced with a
+replacement character if allowed.
+
+=item *
+
+C<allow_noncontinuation> - allow non-continuation bytes where
+continuation bytes are expected.  This is an encoding error and is
+replaced with a replacement character if allowed.
+
+
+=item *
+
+C<strict> - disallows all encoding errors, non-characters, surrogates
```

Comment:
non-characters are now allowed in unicode; maybe strict should allow them? http://www.unicode.org/versions/corrigendum9.html

Reply:
AFAIK they were always allowed, but the current defaults are one of the few things retained from the code I started from.